cranelift/interpreter/src/step.rs

```diff
@@ -557,7 +557,30 @@ where
             false,
         )?),
         Opcode::Ineg => binary(Value::sub, Value::int(0, ctrl_ty)?, arg(0)?)?,
-        Opcode::Iabs => unimplemented!("Iabs"),
+        Opcode::Iabs => {
+            let (min_val, _) = ctrl_ty.lane_type().bounds(true);
+            let min_val: V = Value::int(min_val as i128, ctrl_ty.lane_type())?;
+            if ctrl_ty.is_vector() {
+                let arg0 = extractlanes(&arg(0)?, ctrl_ty.lane_type())?;
+                let new_vec = arg0
+                    .into_iter()
+                    .map(|lane| {
+                        if Value::eq(&lane, &min_val)? {
+                            Ok(min_val.clone())
+                        } else {
+                            Value::int(lane.into_int()?.abs(), ctrl_ty.lane_type())
+                        }
+                    })
+                    .collect::<ValueResult<SimdVec<V>>>()?;
+                assign(vectorizelanes(&new_vec, ctrl_ty)?)
+            } else {
+                assign(if Value::eq(&arg(0)?, &min_val)? {
+                    min_val.clone()
+                } else {
+                    Value::int(arg(0)?.into_int()?.abs(), ctrl_ty.lane_type())?
+                })
```

Comment:
Would it be an idea to make extractlanes and vectorizelanes on non-vector values produce and accept a single element? That way you can use the same code for scalar and vector types without any duplication.

Reply:
Would this make sense as a separate patch? Since I think that would involve restructuring both `extractlanes` and `vectorizelanes` a fair bit, plus then going through the existing uses of those functions in the rest of `step.rs`.

Reply:
This sounds like a reasonable refactor; I also agree that it's a bit out-of-scope for this PR, but could be a followup.